alphagov/paas-admin
# [WIP] Introduce API tokens

What
----

We’d like to have the ability of modification permissions for users that
are not necessarily humans. PaaS Accounts requirement is restricting
this from happening.

We’re still logging, whilst not throwing any errors.

Users started to come into the support, asking for non-email based users
which could be used for their CI systems to interact with PaaS.

This feature, is simply based on the username and password system, but
relies on specific user naming, in order to differentiate between the
Users and Tokens.

An Admin and Org Manager should be able to create unlimited number of
these tokens, retreive the Key and Secret, setup the permissions for
Robot User and use the feature as they would with a google group user.

Now that we are able to have API Tokens, we don't necessarily want to
display them on the same list as the users. This could be confusing.
Filtering out by name should allow us to have the correct numbers in
place.

[WIP]

It would be great if the feature allowed for easy rotation of the
secret.

At the moment, this could be achieved by deleting the user and
re-creating. This however could be problematic when one would want to
replicate the permissions of the deleted user - especially when revoking
the token and not noting what access it used to have.

This feature however, would require either an Admin token, or the Pazmin
tool to have `uaa.admin` permissions, which it currently doesn't have.

How to review
-------------

- Sanity check the idea
- Deploy to your environment
- Attempt to use the feature
  - Create token
  - Rotate token
  - Revoke token
  - List users - see no tokens
  - List tokens - see no users
